File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int totalProd = 1;
        boolean hasZero = false;
        for(int i = 0; i < nums.length; i++){
            if(!hasZero){
                if(nums[i]==0){
                    hasZero = true;
                }else{
                    totalProd *= nums[i];
                }
            }else{
                totalProd *= nums[i];
            }
        }
        
        
        for(int i = 0; i < nums.length; i++){
            
            if(hasZero){
                if(nums[i] != 0){
                    nums[i] = 0;
                }else{
                    nums[i] = totalProd;
                }
            }else{
                nums[i] = totalProd/nums[i];
            }
        }
        return nums;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        long long prod = 1;
        int zeroes = 0;
        for(int i = 0; i < n; i++){
            if(nums[i] != 0) prod *= nums[i];
            if(nums[i] == 0) zeroes++;
        }
        vector<int> ans(n, 0);
        if(zeroes == 0){
            for(int i = 0; i < n; i++){
                ans[i] = prod/nums[i];
            }
        }else if(zeroes == 1){
            for(int i = 0; i < n; i++){
                if(nums[i] == 0){
                    ans[i] = prod;
                    return ans;
                }
            }
        }else if(zeroes > 1) {
            return ans;
        }
        
        return ans;
    }
};